### PR TITLE
AWS DC: Try to remove ELB if NLB doesn't exist

### PR DESCRIPTION
**Pull request by drewgonzales360**
# Summary

If Nelson loadbalancers were provisioned with an older version of Nelson, classic AWS loadbalancers were created. This PR makes deleting those loadbalancers possible with the newer versions of Nelson IE 0.15.x

# Tested
I hopped into the repl with a few scripts and checked out a commit of Nelson that created classic loadbalancers. I created a classic loadbalancer from Nelson.

Then I checked out the head of this branch which uses NLB's. I called loadbalancers.delete and watched Nelson fail to delete the NLB, then try to delete the ELB and succeed. 

```
01:15:12.394 [run-main-1] ERROR nelson.loadbalancers.Aws - aws call to get loadbalancer ARN classic--1-0-0--hashhash failed.
01:15:12.395 [run-main-1] INFO  nelson.loadbalancers.Aws - couldn't delete NLB because Loadbalancer deployment (could not find loadbalancer and therefore could not delete loadbalancer) failed because Load balancers '[classic--1-0-0--hashhash]' not found (Service: AmazonElasticLoadBalancing; Status Code: 400; Error Code: LoadBalancerNotFound; Request ID: adsfsadfasdf-asdfasdf-asdfasdfsadf). attempting to delete ELB
```

I also tested creating and deleting NLB's

## Our Current Use Case
We used to use a version of Nelson that provisioned classic loadbalancers. To use the "Network Loadbalancers", we upgraded the java sdk to use the loadbalancingv2 package. This package can't delete "classic loadbalancers" ie it is not backwards compatible.

### CODE REVIEW

File: core/src/main/scala/loadbalancers/Aws.scala [hunk 1]
```diff
@@ -66,28 +66,33 @@ final class Aws(cfg: Infrastructure.Aws) extends (LoadbalancerOp ~> IO) {
...
+      _ <- deleteNLB(lbarn)
       _ <- deleteTargetGroups(targetGroups)
 
       _ <- deleteASG(name)
-    } yield ()
+    } yield ()).recoverWith{
+      case e: Exception =>
+        log.info(s"couldn't delete NLB because ${e.getMessage}. attempting to delete ELB")
+        for {
+          _ <- deleteELB(name)
```
Comment (timperrett) [inline core/src/main/scala/loadbalancers/Aws.scala]: What happens if these fail?

Comment (drewgonzales360) [reply core/src/main/scala/loadbalancers/Aws.scala]: Errors will get logged, but Nelson will still mark the resources as terminated. Not much we can do :man_shrugging:

**Comment by drewgonzales360** [comment]
@lu4nm3 @timperrett

File: core/src/main/scala/loadbalancers/Aws.scala [hunk 2]
```diff
@@ -97,35 +92,82 @@ final class Aws(cfg: Infrastructure.Aws) extends (LoadbalancerOp ~> IO) {
...
       r <- IO(cfg.nlb.describeLoadBalancers(describeLoadBalancerRequest))
       x = r.getLoadBalancers.asScala.toList
       a = x.head.getLoadBalancerArn
-    } yield a).recoverWith{
+    } yield Option(a)).recoverWith{
       case e: Exception =>
-        log.error(s"aws call to get loadbalancer ARN ${describeLoadBalancerRequest.getNames.asScala.head} failed.")
-        IO.raiseError(FailedLoadbalancerDeploy(s"could not find loadbalancer and therefore could not delete loadbalancer", e.getMessage))
+        log.error(s"aws call to get loadbalancer ARN ${describeLoadBalancerRequest.getNames.asScala.head} failed with ${e.getMessage}")
+        IO(None)
```
Comment (timperrett) [inline core/src/main/scala/loadbalancers/Aws.scala]: Does it semantically make sense to have an `Option` return here? It's unclear to me what successful result scenario would be `None`.

Comment (drewgonzales360) [reply core/src/main/scala/loadbalancers/Aws.scala]: Yeah I have no idea. If the network call is successful, then I want to return the string. If it's not successful, I'm not sure what to return.

In my mind, I was thinking that I want to return _something_ if the call is successful and _nothing_ if it's not. `Option` seemed like the way to go. What did you have in mind?

File: core/src/main/scala/loadbalancers/Aws.scala [hunk 3]
```diff
@@ -97,35 +92,82 @@ final class Aws(cfg: Infrastructure.Aws) extends (LoadbalancerOp ~> IO) {
   // The target groups have a restriction such that name of the target group can't
   // have "double dashes" and the entire name must be less than 32 characters.
   // Unfortunately, target groups will not have the same name as the loadbalancers.
-  // An example target group will be named stack-1-0-0-443 where the last number is the
-  // port that it listens on (instead of the hash). This allows stack-1-0-0-80 to be
+  // An example target group will be named stack-aevaif-443 where the last number is the
```
Comment (timperrett) [inline core/src/main/scala/loadbalancers/Aws.scala]: Why would the revision not be a part of the name?

File: core/src/main/scala/loadbalancers/Aws.scala (hunk 2 above)
Comment (timperrett) [reply core/src/main/scala/loadbalancers/Aws.scala]: In this case, one would not usually leverage `Option`; here we would use `Either` where the result is _either_ a `String` or `Error`, but the resulting type is clear in either branch (semantically, and practically). In essence, this is what you're getting with `IO` - your computation either succeeds or it doesn't, but the semantics of IO are specifically for IO-related things (which you have in this case).

Does that help?

**Comment by drewgonzales360** [comment]
Closing this in favor of another PR. I have a private fork where this feature is working.

**Pull request closed without merging**